File: src/reo-core/org.ect.reo/src/org/ect/reo/channels/SyncSpout.java
/*******************************************************************************
 * <copyright>
 * This file is part of the Extensible Coordination Tools (ECT).
 * Copyright (c) 2013 ECT developers. 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * </copyright>
 *******************************************************************************/
package org.ect.reo.channels;

import org.eclipse.emf.ecore.EClass;
import org.ect.reo.Node;
import org.ect.reo.animation.Animation;
import org.ect.reo.animation.AnimationTable;
import org.ect.reo.animation.CreateStep;
import org.ect.reo.animation.SendStep;
import org.ect.reo.colouring.FlowColour;



/**
 * @see org.ect.reo.channels.ChannelsPackage#createSyncSpout()
 * @model kind="class"
 * @generated
 */
public class SyncSpout extends SpoutChannel {

	/**
	 * @see org.ect.reo.channels.SpoutChannel#SpoutChannel(Node, Node)
	 * @generated NOT
	 */
	public SyncSpout(Node node1, Node node2) {
		super(node1, node2);
	}
	
	/**
	 * Check whether this channel is synchronous.
	 * @generated NOT
	 */
	@Override
	public boolean isSynchronous() {
		return true;
	}

	/**
	 * @generated NOT
	 */
	public AnimationTable getAnimationTable() {

		AnimationTable table = new AnimationTable();
		Animation flow, noflow1, noflow2;
		
		if (getSinkEndOne()==null || getSinkEndTwo()==null) {
			return table;
		}

		// Flow.
		flow = new Animation();
		flow.setColour(getSinkEndOne(), FlowColour.FLOW_LITERAL);
		flow.setColour(getSinkEndTwo(), FlowColour.FLOW_LITERAL);
		flow.setNextAnimationTable(table);
		
		CreateStep createStep = new CreateStep();
		createStep.getActiveEnds().add(getSinkEndOne());
		createStep.getActiveEnds().add(getSinkEndTwo());
		
		SendStep sendStep = new SendStep();
		sendStep.getActiveEnds().add(getSinkEndOne());
		sendStep.getActiveEnds().add(getSinkEndTwo());
		
		flow.getSteps().add(createStep);
		flow.getSteps().add(sendStep);

		table.add(flow);
		
		// No flow -<-<-
		noflow1 = new Animation();
		noflow1.setColour(getSinkEndOne(), FlowColour.NO_FLOW_GIVE_REASON_LITERAL);
		noflow1.setColour(getSinkEndTwo(), FlowColour.NO_FLOW_REQUIRE_REASON_LITERAL);
		noflow1.setNextAnimationTable(table);
		
		table.add(noflow1);

		
		// No flow ->->-
		noflow2 = new Animation();
		noflow2.setColour(getSinkEndOne(), FlowColour.NO_FLOW_REQUIRE_REASON_LITERAL);
		noflow2.setColour(getSinkEndTwo(), FlowColour.NO_FLOW_GIVE_REASON_LITERAL);
		noflow2.setNextAnimationTable(table);

		table.add(noflow2);

		return table;
				
	}

	
	/* ---------------------------------------------------------------- *
	 * GENERATED CODE.                                                  *
	 * Do not edit below this line. If you need to edit, move it above  *
	 * this line and change the '@generated'-tag to '@generated NOT'.   *
	 * ---------------------------------------------------------------- */

	/**
	 * @generated
	 */
	public SyncSpout() {
		super();
	}
	
	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return ChannelsPackage.Literals.SYNC_SPOUT;
	}

}

File: src/reo-core/org.ect.reo.distengine/src/org/ect/reo/distengine/colouring/primitives/SyncSpout.scala
/*
 * <copyright>
 * This file is part of the Extensible Coordination Tools (ECT).
 * Copyright (c) 2013 ECT developers. 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * </copyright>
 */
package org.ect.reo.distengine.colouring.primitives;

//import org.ect.reo.distengine.colouring._;
//import org.ect.reo.distengine.colouring.
import org.ect.reo.distengine.common._;
import org.ect.reo.distengine.common.Utils._;
import org.ect.reo.distengine.redrum.{Primitive, End}
import collection.mutable.HashMap
import collection.mutable.HashSet
import scala.collection.Set
import org.ect.reo.distengine.colouring._


/**
   * Definition of a synchronous spout channel, using the semantics
   * provided by <code>org.ect.reo.distengine.colouring.CC</code>.
   */
class SyncSpout(prim:PrimID,logger:Logger) extends Primitive[org.ect.reo.distengine.colouring.CC](logger) {
  val name = prim
  val endl = prim + "-l"
  val endr = prim + "-r"
  val ends : HashMap[EndLID,AbstractEnd] = new HashMap()
  
  // initial behaviour with no CT, so the ends can be initiallised, required to define the CT
  val behaviour = new org.ect.reo.distengine.colouring.CC(new ColouringTable)
  
  val a : AbstractEnd = new End(endl,Output,this)
  val b : AbstractEnd = new End(endr,Output,this)
  val auid = EndUID(prim,endl)
  val buid = EndUID(prim,endr)


  def source(ct:Colouring,e:EndUID) : DataSource = {
    return new Right(null)
  }
  
  def updateStep(c:Colouring) = {}
  
  ends(endl) = a
  ends(endr) = b
  
  // the ColTable of the sspout channel
  val col1 = new Colouring
  col1(auid) = Flow
  col1(buid) = Flow
  behaviour.value += col1
  val col2 = new Colouring
  col2(auid) = FromOut
  col2(buid) = ToOut
  behaviour.value += col2
  val col3 = new Colouring
  col3(auid) = ToOut
  col3(buid) = FromOut
  behaviour.value += col3  
  
}
